clusterspec/clusterspec.go

```diff
@@ -0,0 +1,133 @@
+package clusterspec
+
+import (
+	"net/url"
+	"strings"
+
+	v1alpha1core "github.com/giantswarm/apiextensions/pkg/apis/core/v1alpha1"
+	v1alpha1provider "github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1"
+	"github.com/giantswarm/certs"
+	"github.com/giantswarm/microerror"
+)
+
+// Config holds required configuration values for factory to be able to
+// construct provider independent Cluster configuration based on
+// ClusterGuestConfig.
+type Config struct {
+	Calico struct {
+		CIDR   int
+		MTU    int
+		Subnet string
+	}
+
+	Docker struct {
+		Daemon struct {
+			CIDR      string
+			ExtraArgs string
+		}
+	}
+
+	Etcd struct {
+		AltNames string
+		Port     int
+		Prefix   string
+	}
+
+	Kubernetes struct {
+		API struct {
+			AltNames       string
+			ClusterIPRange string
+			InsecurePort   int
+			SecurePort     int
+		}
+
+		Domain string
+
+		Hyperkube struct {
+			Docker struct {
+				Image string
+			}
+		}
+
+		IngressController struct {
+			BaseDomain string
+			Docker     struct {
+				Image string
+			}
+			InsecurePort int
+			SecurePort   int
+		}
+
+		Kubelet struct {
+			AltNames string
+			Labels   string
+			Port     int
+		}
+
+		NetworkSetup struct {
+			Docker struct {
+				Image string
+			}
+		}
+
+		SSH struct {
+			UserList []struct {
+				Name      string
+				PublicKey string
+			}
+		}
+	}
+
+	Provider struct {
+		Kind string
+	}
+}
+
+// Factory to construct provider independent cluster specifications
+type Factory interface {
+	// New constructs provider/v1alpha1.Cluster
+	New(clusterGuestConfig v1alpha1core.ClusterGuestConfig) (*v1alpha1provider.Cluster, error)
+}
+
+type factory struct {
+	config *Config
+}
+
+// NewFactory constructs new Factory
+func NewFactory(c *Config) (Factory, error) {
```

Comment:
Do we really need to specify that this function returns an interface type? I think it would make more sense in the functions that receive values implementing that interface, WDYT?

Reply:
Do you mean having something like this:
```go
type ClusterConfig struct {}

func Build(clusterGuestConfig v1alpha1core.ClusterGuestConfig, cfg *ClusterConfig) (v1alpha1provider.Cluster, error) {}
```

I think this could work. Only minor thing here is that now component that doesn't need to know anything about how `Cluster` is constructed, have access to configuration that is used for its construction. This breaks encapsulation I think. WDYT?

Reply:
No, the interface is ok, I was wondering if we need to declare that `NewFactory` returns a `Factory` and, instead, just in the functions that are going to use it, declare that they need a `Factory`. That would require exporting the implementing type though, and that type is only meaningful because of it is implementing the interface (it doesn't have any additional method attached), so on second thought I like this approach more, didn't notice before that the implementing type is private :+1:

Reply:
I agree with @fgimenez here, I always prefer to return and keep reference to concrete type as long as possible. It makes it easier to analyze the code.

Reply:
@fgimenez Yeah, actually the interface as of itself is not that important here. What I really want is only a factory function, but in order to implement it a supplied configuration is required. Interface is product of my thought process to enforce validation of `Config` before use of factory. See a bit more details below.